### ISSUE TITLE
Display Mapsforge as real Vector Tiles

### ISSUE THREAD

**Issue by monsieurtanuki** [osmdroid/osmdroid#1156]
## Issue Type

- [x] Improvement

## Description and/or steps/code to reproduce the problem

So far, we display Mapsforge tiles as 256x256 tiles, that we stretch if we are on a floating point zoom level (vs. integer). This matches everything we do with 256x256 raster tiles.
Now that we have floating point zoom levels, it would make sense to have Mapsforge display the tiles directly to the right dimension (e.g. 400x400), without stretching a 256x256 bitmap, but directly rendering vectors into the target bitmap.
Doing so we would have a clean display for Mapsforge tiles, as opposed to the current magnified pixel version.

Could be related to #246.

### Version of osmdroid the issue relates to:

6.0.2

**Comment by secaw** [comment]
Would be fantastic. On the other hand mapsforge rendering is very CPU intense, especially the current osmdroid implementation. IMHO the current way we couldn't afford it battery wise at all.
The mapsforge integration would have to be most native in order to avoid CPU overhead and then we would also save storage space, as we'd not use the osmdroid cache anymore.
Is this really possible?

**Comment by monsieurtanuki** [comment]
> mapsforge rendering is very CPU intense, especially the current osmdroid implementation

I wasn't aware of that.
A solution could be to use memory cache very wisely, and to prevent mapsforge vectorial computations during zoom animation.

**Comment by monsieurtanuki** [comment]
@secaw Regarding Mapsforge performances, do you mean in general on Android or more specifically through osmdroid? Because there might be optimization within osmdroid, cf. [MapsForgeTileModuleProvider](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-mapsforge/src/main/java/org/osmdroid/mapsforge/MapsForgeTileModuleProvider.java#L91).

**Comment by secaw** [comment]
Yes, the osmdroid implementation in particular. When using apps based on native mapsforge mapview, map display is way more smooth than it's tile rendering within osmdroid.

**Comment by monsieurtanuki** [comment]
Then the fix I have in mind might work, at least marginally.

**Comment by monsieurtanuki** [comment]
On my smartphone, my fix makes things 75ms faster for each tile on average.

**Comment by secaw** [comment]
That's quite something, isn't it. How does it work?

**Comment by secaw** [comment]
Alright, I always found it awkward, that tiles of an offline map source are being stored again in a different way. This seems like a waste of storage space at least. On the other hand considering the compute resources needed to actually view a vector based map, it makes sense, especially when viewing parts of the map over and over again, which is a common way of using a map. So I think we really need to weigh the pros and cons carefully before getting rid of writing to cache, because if you review a part of the map again and again, using a cached version will still be better battery usage wise, than rerendering it every time. Might need some extensive testing.
Considering this it might be at least a first step to optimize tile writing as you suggest and make it optional. Being optional, would you recommend the default be on or off?

**Comment by secaw** [comment]
Did some more testing with tile writing only of zoom levels 15 and below. I now get a smooth map display after first view everywhere.

**Comment by monsieurtanuki** [comment]
Disclaimer: I'm not a mapsforge user myself, and I only worked with it in the context of #1145 (zoom levels around 14), testing with `MapsforgeTileProviderSample` that uses only the mapsforge provider (and not the `TileWriter` cache as a possible input).

What I can say is:
* it's very interesting to know that the higher the zoom level, the faster the tile rendering is
* I still noticed faster rendering without the `TileWriter` part - could you run tests on zoom level 15 in order to see how many ms faster we are? (or maybe I was just lucky when I ran the tests)
* regarding tile accuracy, as explained in #1145, we cannot rely on mapsforge tile's first version (potential overlapping labels), therefore it means that we need to be able to refresh the tiles. In my case it was just about setting tiles as `EXPIRED` in the _memory_ cache, and it meant something because the "mapsforge overlapping label repository" was in memory too. If we put the tiles into a `TileWriter` cache, we will lose the ability to use meaningfully this repository - therefore potentially we'll get the same "broken labels" bug.
* perhaps a solution would be to support a mapsforge conversion into raster tiles, and then use raster tiles only

**Comment by secaw** [comment]
Ok, did some measurements myself. It is all not very comparable, because as said before, rendering times in mapsforge absolutely depend on the amount of vectors included in the given tile. Tiles from low zoom levels of course include a multiple of vector information compared to high zoom level tiles.
I did notice a comparable increase in display time of about 70ms if the given tile is stored with `TileWriter` like you measured, but for low zoom level tiles this is not the major amount:
Tiles from zoom level 11 take around 500ms average without storing.
Tiles from zoom level 16 take around 50ms average without storing.
Stored tiles from any zoom level when loaded again from `TileWriter` take around 15ms!
So using the `TileWriter` makes a lot of sense, especially for low zoom level tiles that are also more likely to be re-viewed. With this knowledge I will change my custom mapsforge provider now this way, that I will only store tiles up to zoom level 15.
Also as said before, I don't encounter very much of the broken lables anymore. Maybe it's got to do with using mapsforge 0.9.1, or could it be that the new `getTileCache().getProtectedTileContainers()` actually help, because more neighboring tiles are computed at once?
Anyway, I still think, that we can only get rid of the `TileWriter` mechanism in osmdroid, if mapsforge is being integrated in a more native way, as rendering in mapsforge's own mapview is absolutely smooth.

**Comment by monsieurtanuki** [comment]
@secaw I won't be able to work on this issue anymore as I had to change my priorities. Sorry!

**Comment by secaw** [comment]
@monsieurtanuki Alright, thanks a lot anyways! Will it make sense to use your new mapsforge direct renderer in the future?

**Comment by monsieurtanuki** [comment]
@secaw I guess so, but I wasn't in the best configurations to test it completely. Anyway, it seems to have fixed #1145, and regarding osmdroid, we need to wait for mapsforge's next release in order to close #1145.

For the record, the attached version of `DirectRenderer` includes the "out-of-box tile returns NULL" feature but I couldn't merge it. Feel free to merge it into mapsforge if needed.
[DirectRenderer.zip](https://github.com/osmdroid/osmdroid/files/2449766/DirectRenderer.zip)

**Comment by secaw** [comment]
Thanks a lot @monsieurtanuki, I will check it out.
